### PR TITLE
First implementation of Event system

### LINKED ISSUE #36: DOM Events

### PR DESCRIPTION
**Pull request by francisbouvier**
Closes #36

### CODE REVIEW

File: src/dom/event_target.zig [hunk 1]
```diff
@@ -6,4 +13,55 @@ pub const EventTarget = struct {
...
+        fn handle(event: ?*parser.Event, data: ?*anyopaque) callconv(.C) void {
+            const ptr: *align(@alignOf(*Callback)) anyopaque = @alignCast(data.?);
+            const func = @as(*Callback, @ptrCast(ptr));
+            func.call(.{event}) catch unreachable;
+            // NOTE: we can not call func.deinit here
+            // b/c the handler can be called several times
+            // as the event goes through the ancestors
+            // TODO: check the event phase to call func.deinit and free func
+        }
+    }.handle;
```
Comment (krichprollsch) [inline src/dom/event_target.zig]: I'm wondering if this struct shouldn't be moved to `netsurf.zig` instead.
Than we can hide plumbing from here.

File: src/dom/event_target.zig [hunk 2]
```diff
@@ -6,4 +13,55 @@ pub const EventTarget = struct {
...
+    pub fn _addEventListener(
+        self: *parser.EventTarget,
+        alloc: std.mem.Allocator,
+        eventType: []const u8,
+        cbk: Callback,
+    ) !void {
+        // TODO: when can we free this allocation?
+        const cbk_ptr = try alloc.create(Callback);
+        cbk_ptr.* = cbk;
+        const func = @as(*anyopaque, @ptrCast(cbk_ptr));
```
Comment (krichprollsch) [inline src/dom/event_target.zig]: Same here, I think we could move the callback casting into `eventTargetAddEventListener` and call
```zig
 try parser.eventTargetAddEventListener(self, eventType, cbk_ptr);
 ```

File: src/dom/event_target.zig [hunk 3]
```diff
@@ -6,4 +13,55 @@ pub const EventTarget = struct {
...
+pub fn testExecFn(
+    _: std.mem.Allocator,
+    js_env: *jsruntime.Env,
+) anyerror!void {
+    var basic = [_]Case{
+        .{ .src = "let event = new Event('myEvent')", .ex = "undefined" },
+        .{ .src = "let content = document.getElementById('content')", .ex = "undefined" },
+        .{ .src = "var nb = 0; content.addEventListener('myEvent', function(event) {nb ++;})", .ex = "undefined" },
+        .{ .src = "content.dispatchEvent(event)", .ex = "true" },
+        .{ .src = "nb", .ex = "2" }, // 2 because the callback is called twice
```
Comment (krichprollsch) [inline src/dom/event_target.zig]: The same code in chrome generates `1`

File: src/dom/event_target.zig [hunk 4]
```diff
@@ -6,4 +13,55 @@ pub const EventTarget = struct {
...
+        }
+    }.handle;
+
+    pub fn _addEventListener(
+        self: *parser.EventTarget,
+        alloc: std.mem.Allocator,
+        eventType: []const u8,
+        cbk: Callback,
+    ) !void {
+        // TODO: when can we free this allocation?
```
Comment (krichprollsch) [inline src/dom/event_target.zig]: on the removeListener call maybe?

File: src/dom/event_target.zig (hunk 1 above)
Comment (francisbouvier) [reply src/dom/event_target.zig]: Yes I was asking myself the same question.
I think you are right, it's better on netsurf.

File: src/dom/event_target.zig (hunk 4 above)
Comment (francisbouvier) [reply src/dom/event_target.zig]: Yes of course. But I want to investigate the lifetime of the Callback to see if we can do it before.

File: src/dom/event_target.zig (hunk 3 above)
Comment (francisbouvier) [reply src/dom/event_target.zig]: Fixed

**Comment by krichprollsch** [comment]
In order to try to pass some of the new WPT tests, I suggest to implement `document.createEvent` https://dom.spec.whatwg.org/#dom-document-createevent https://developer.mozilla.org/en-US/docs/Web/API/Document/createEvent


+ `initEvent`

File: src/netsurf.zig [hunk 5]
```diff
@@ -451,13 +451,58 @@ const event_handler = struct {
     }
 }.handle;
 
+// EventListener
+pub const EventListener = c.dom_event_listener;
+
+pub fn eventListenerGetData(lst: *EventListener) ?*anyopaque {
```
Comment (krichprollsch) [inline src/netsurf.zig]: You could directly return a `*Callback` here isn't it?

File: src/netsurf.zig [hunk 6]
```diff
@@ -451,13 +451,58 @@ const event_handler = struct {
...
+}
+
 // EventTarget
 pub const EventTarget = c.dom_event_target;
 
 fn eventTargetVtable(et: *EventTarget) c.dom_event_target_vtable {
     return getVtable(c.dom_event_target_vtable, EventTarget, et);
 }
 
+pub fn eventTargetHasListener(et: *EventTarget, typ: []const u8, cbk_id: usize) !?*EventListener {
```
Comment (krichprollsch) [inline src/netsurf.zig]: I'm not sure about the `cbk_id: usize`.
It seems inconsistent w/ the `eventTargetAddEventListener` signature which expects a `cbk_ptr: *Callback`.

**Pull request merged**